File: asg_5/hgle_asg5.py
# hgle _asg5.py
# program for asg5 methods spring 2026
# 2026-03-15
def main():

    print("Hi and welcome to the BMI population analyzer!")

    units = input("Please choose Metric or Imperial to begin: ")

    if units == "Metric":
      # ask user for a metric weight value 
      weight = float(input("Give me a weight in kilograms: "))

      # ask user for a metric height value 
      height = float(input("Give me a height in cm: "))

      # convert cm height to m height
      height = height/100

      # calculate bmi using metric units
      bmi_val = (weight / height ** 2) 

      # prints bmi value
      print(f"bmi is {bmi_val}")
    elif units == "Imperial":
      # ask user for a imperial weight value 
      weight = float(input("Give me a weight in pounds: "))

      # ask user for a imperial height value 
      height = float(input("Give me a height in inches: "))

      # calculate bmi using imperial units
      bmi_val = (weight / height ** 2) * 703

      # prints bmi value
      print(f"bmi is {bmi_val}")
    
    # determine weight status
    if bmi_val >= 40.0:
        upper_bound = float('inf')
        lower_bound = 40.0

        weight_status = "extreme or high risk obesity"
    elif bmi_val <= 39.9 and bmi_val > 30.0:
        upper_bound = 40.0
        lower_bound = 30.0

        weight_status = "obese"
    elif bmi_val <= 30.0 and bmi_val > 25.0:
        upper_bound = 30.0
        lower_bound = 25.0

        weight_status = "overweight"
    elif bmi_val <= 25.0 and bmi_val > 18.5:
        upper_bound = 25.0
        lower_bound = 18.5

        weight_status = "normal"
    else: 
        upper_bound = 18.5
        lower_bound = -float('inf')

        weight_status = "underweight"
    
    # prints weight status
    print(f"weight status is {weight_status}")

    # define file names
    measurement_file_name    = "/oscar/data/shared/ursa/synthetic_ri/demo_omop/measurement.csv"
    detail_output_file_name  = "hgle_asg5_detail_results.txt"
    summary_output_file_name = "hgle_asg5_summary_results.txt"

    # create file handles
    measurement_file    = open(measurement_file_name,    "r")
    detail_output_file  = open(detail_output_file_name,  "w")
    summary_output_file = open(summary_output_file_name, "a")

    # create counters
    hi_bmi_count = 0
    in_bmi_count = 0

    # read file
    for line in measurement_file:

        line = line.rstrip("\n")
        
        # split line by comma
        line_part_array = line.split(",")
        
        # pull pt id and concept id
        pt_id           = line_part_array[1]
        omop_concept_id = line_part_array[2]

        # process lines that have BMI (concept_id 3038553)
        if omop_concept_id == "3038553":

            # retrieve value as a Float64
            value = float(line_part_array[8])

            # increment appropriate counter
            # and print out appropriate value to detail file
            if value >= upper_bound:
                hi_bmi_count += 1
                detail_output_file.write(f"{pt_id}|{value}|HI\n")
            elif value >= lower_bound and value <= upper_bound:
                in_bmi_count += 1
                detail_output_file.write(f"{pt_id}|{value}|IN\n")

    #print out summary
    print(f"{bmi_val}|{lower_bound}-{upper_bound}|{in_bmi_count}|{hi_bmi_count}")

    summary_output_file.write(f"{bmi_val}|{lower_bound}-{upper_bound}|{in_bmi_count}|{hi_bmi_count}\n")

    # close 
    measurement_file.close()
    detail_output_file.close()
    summary_output_file.close()

# call main function
main()

File: asg_5/hgle_asg5.jl
# hgle_asg5.jl
# program for asg5 methods spring 2026
# 2026-03-15

function main()

    # greet user
    println("hello! I will count race for type 2 diabetus mellitus")

    # define file names
    condition_file_name = "/oscar/data/shared/ursa/synthetic_ri/demo_omop/condition_occurrence.csv"
    person_file_name    = "/oscar/data/shared/ursa/synthetic_ri/demo_omop/person.csv"
    output_file_name    = "hgle_asg5_output.txt"

    # create file handles 
    condition_file    = open(condition_file_name, "r")
    person_file       = open(person_file_name, "r")
    output_count_file = open(output_file_name, "w")

    # define dx code for T2DM 
    dx_code = ["201826", "192279"]

    # create pt_id_set
    pt_id_set = Set()

    println("I am going to look up patients who have Dx codes $(dx_code[1]) or $(dx_code[2])")

    # read throught the condition file and store patient ids for where they have the Dx
    for condition_file_line in eachline(condition_file)
        
        # split line into an ARRAY
        line_part_array = split(condition_file_line, ",")

        condition_code = line_part_array[3]
        pt_id          = line_part_array[2]

        # if the condition code mathces the dx_code,
        # push the patient id into the set
        if condition_code in dx_code

            push!(pt_id_set, pt_id)

        end

    end

    #println(pt_id_set)
    number_patients = length(pt_id_set)
    println("I found $number_patients patients with T2DM!!!!!")

    # create lookup dictionaries
    race_dict = Dict()

    # add lookip values to dictionaries
    race_dict["8515"] = "Asian"
    race_dict["8516"] = "Black or African American"
    race_dict["8527"] = "White"
    race_dict["0"]    = "Unknown"

    race_count_dict = Dict()

    # read through the person file
    for person_file_line in eachline(person_file)
        
        # split the line 
        line_part_array = split(person_file_line, ",")

        # get pt id
        pt_id = line_part_array[1]
        
        if in(pt_id, pt_id_set)

            # pull the race codes
            # println(line_part_array[7])
            race_code = line_part_array[7]

            println("for patient $pt_id, the race_code is $race_code")

            race_string = race_dict[race_code]

            println("for patient $pt_id, the race_string is $race_string")
            
            # increment race dictionary counter accordingly
            if haskey(race_count_dict, race_string)
                race_count_dict[race_string] +=1
            else
                race_count_dict[race_string] = 1
            end

        end

    end

    for race in keys(race_count_dict)
        output_string = "$race|$(race_count_dict[race])\n"
        print(output_string)
        print(output_count_file, output_string)
    end

end

main()